src/Shouldly/ShouldlyExtensionMethods/ShouldBeNullExtensions.cs

```diff
@@ -26,19 +26,17 @@ public static void ShouldBeNull<T>([MaybeNull] this T? actual, string? customMes
         }
 
         [ContractAnnotation("actual:null => halt")]
-        public static void ShouldNotBeNull<T>([NotNull] this T? actual, string? customMessage = null)
+        public static T ShouldNotBeNull<T>([NotNull] this T? actual, string? customMessage = null)
```

Comment:
Don't you want to return `T?` instead of `T`, even if it has the `NotNullAttribute` and that it will throw if the value is `null`?

It seems weird that a fluent method is not returning the same type as its parameter, meaning that the chained assertion will be performed on a different type.

For example, this won't compile : 
`((int?)0).ShouldNotBeNull().ShouldNotBeNull();`

Reply:
Thanks for the question! No, the motivation in #559 was specifically to unwrap nullable values. I have a lot of code asserts that a struct is not null and then asserts things about several deeper properties. Having to repeat `.Value` is onerous. There are code examples in #559 that show examples.

(This is also why we petitioned for C# 8's coalescing assignments to unwrap nullable values. The original design did not; it made you do `(nullableFoo ??= GetFoo()).Value`. Having to say `.Value` on a thing that is never null is noisy and frustrating. Thankfully, they listened and changed it before C# 8 shipped!)

I see it as a feature that `((int?)0).ShouldNotBeNull().ShouldNotBeNull();` doesn't compile, in line with https://github.com/shouldly/shouldly/issues/584, and I like the parallel with nullable reference types.

Reply:
Another longstanding example of an assertion method that returns a different type is ShouldHaveSingleItem.